File: 1201-1300/1293-Shortest Path in a Grid with Obstacles Elimination/1293-Shortest Path in a Grid with Obstacles Elimination.py
import collections
class Solution:
    def shortestPath(self, grid: List[List[int]], k: int) -> int:
        m, n = len(grid), len(grid[0])
        if k >= m - 1 + n - 1:
            return m + n - 2
        Q = collections.deque([(0, 0, k)])
        visited = set([(0, 0, k)])
        step = 0
        while Q:
            for _ in range(len(Q)):
                r, c, e = Q.popleft()
                if r == m - 1 and c == n - 1:
                    return step
                for nr, nc in (r-1, c), (r+1, c), (r, c-1), (r, c+1):
                    if 0 <= nr < m and 0 <= nc < n:
                        if grid[nr][nc] == 0 and (nr, nc, e) not in visited:
                            Q.append((nr, nc, e))
                            visited.add((nr, nc, e))
                        elif grid[nr][nc] == 1 and e > 0 and (nr, nc, e - 1) not in visited:
                            Q.append((nr, nc, e - 1))
                            visited.add((nr, nc, e - 1))                
            step += 1
            
        return -1

File: 1201-1300/1293-Shortest Path in a Grid with Obstacles Elimination/1293-Shortest Path in a Grid with Obstacles Elimination.cpp
class Solution {
public:
    int shortestPath(vector<vector<int>>& grid, int k) {
        int m = grid.size(), n = grid[0].size();
        queue<tuple<int, int, int>> Q;
        Q.emplace(0, 0, k);
        int visited[m][n][1 + k];
        memset(visited, 0, sizeof(visited));
        visited[0][0][k] = 1;
        int step = 0;
        int dr[4] = {-1, 1, 0, 0};
        int dc[4] = {0, 0, -1, 1};
        while (!Q.empty()) {
            int qs = Q.size();
            for (int i = 0; i < qs; ++i) {
                auto [r, c, e] = Q.front();
                Q.pop();
                if (r == m - 1 && c == n - 1) {
                    return step;
                }
                for (int k = 0; k < 4; ++k) {
                    int nr = r + dr[k];
                    int nc = c + dc[k];
                    if (nr >= 0 && nr < m && nc >= 0 && nc < n) {
                        if (grid[nr][nc] == 0 && !visited[nr][nc][e]) {
                            Q.emplace(nr, nc, e);
                            visited[nr][nc][e] = 1;
                        }
                        else if (grid[nr][nc] == 1 && e > 0 && !visited[nr][nc][e - 1]) {
                            Q.emplace(nr, nc, e - 1);
                            visited[nr][nc][e - 1] = 1;
                        }
                    }
                }
            }
            ++step;
        }
        
        return -1;
    }
};
